apps/prairielearn/src/pages/administratorFeatures/administratorFeatures.html.ts

```diff
@@ -166,3 +180,105 @@ function FeatureGrant({ featureGrant }: { featureGrant: FeatureGrantRow }) {
     </div>
   `;
 }
+
+interface FeatureGrantModalProps {
+  feature: string;
+  institutions: Institution[];
+  institution_id?: string | null;
+  courses?: Course[];
+  course_id?: string | null;
+  course_instances?: CourseInstance[];
+  course_instance_id?: string | null;
+  csrfToken: string;
+}
+
+function AddFeatureGrantModal(props: FeatureGrantModalProps) {
+  return Modal({
+    title: 'Grant feature',
+    id: 'add-feature-grant-modal',
+    body: AddFeatureGrantModalBody(props),
+    footer: html`
+      <input type="hidden" name="__csrf_token" value="${props.csrfToken}" />
+      <button type="button" class="btn btn-secondary" data-dismiss="modal">Cancel</button>
+      <button type="submit" class="btn btn-primary">Grant feature</button>
+    `,
+  });
+}
+
+export function AddFeatureGrantModalBody({
+  feature,
+  institutions,
+  institution_id,
+  courses,
+  course_id,
+  course_instances,
+  course_instance_id,
+}: Omit<FeatureGrantModalProps, 'csrfToken'>) {
+  const modalUrl = `/pl/administrator/features/${feature}/modal`;
+  return html`
+    <div
+      hx-get="${modalUrl}"
+      hx-trigger="change"
+      hx-target="closest .modal-body"
+      hx-include="closest .modal-body"
+    >
```

Comment:
Very slick!

Is there any way to get htmx to display some kind of loading indicator? This dialog felt awesome on a fast connection but when I simulated 3G it was pretty laggy. I don't think this is really a problem in practice and I don't want to hold up this PR but I'm just curious what the right thing to do is while loading, assuming we are going to use this kind of htmx pattern more often in the future.

Reply:
See 3d3508e - htmx comes with a few extensions that make it *relatively* easy to handle loading states, I tried out one of them and it works reasonably well! The only compromise really is that it currently shows a loading spinner for *all* the dropdowns even when one isn't technically loading. I felt that was most consistent with disabling all the inputs though (which this now also does).

An alternative would be to only show an input when the parent is selected (that is, completely hide the courses `<select>` if institutions is `None`. That might not be horrible to implement? 🤷

None of this really matters for this admin-facing page, of course, but I agree it's nice to be establishing patterns and figuring out what's possible with htmx!

Reply:
Very nice! I love the UX and it's surprisingly little additional markup.